src/tool_operate.c

```diff
@@ -686,19 +686,210 @@ static void single_transfer_cleanup(struct OperationConfig *config)
+    }
+  }
+
+  return NULL;
+}
+
+/*
+ * Rewrite ipfs://<cid> and ipns://<cid> to be a normal
+ * URL that can be handled by an IPFS gateway.
+ */
+static CURLcode ipfs_url_rewrite(CURLU *uh, char *protocol, char **url,
+                                 struct OperationConfig *config)
+{
+  CURLcode result = CURLE_URL_MALFORMAT;
+  char *gateway = NULL;
+  char *cid = NULL;
+  char *pathbuffer = NULL;
+  CURLU *ipfsurl = curl_url();
+
+  CURLUcode ret = curl_url_get(uh, CURLUPART_HOST, &cid, CURLU_URLDECODE);
+  if(ret != CURLUE_OK) {
+    goto clean;
+  }
+
+  if(!cid) {
+    goto clean;
+  }
+
+  /* We might have a --ipfs-gateway argument. Check it first and use it. Error
+   * if we do have something but if it's an invalid url.
+   */
+  if(config->ipfs_gateway) {
+    if(curl_url_set(ipfsurl, CURLUPART_URL, config->ipfs_gateway, 0)
+                    == CURLUE_OK) {
+        gateway = config->ipfs_gateway;
+    }
+    else {
+      result = CURLE_BAD_FUNCTION_ARGUMENT;
+      goto clean;
+    }
+  }
+  else {
+    gateway = ipfs_gateway();
+    if(!gateway) {
+      result = CURLE_FILE_COULDNT_READ_FILE;
+      goto clean;
+    }
+
+    if(curl_url_set(ipfsurl, CURLUPART_URL, gateway, CURLU_GUESS_SCHEME
+                    | CURLU_NON_SUPPORT_SCHEME) != CURLUE_OK) {
+      goto clean;
+    }
+  }
+
+  pathbuffer = aprintf("%s/%s", protocol, cid);
+  if(!pathbuffer) {
+    goto clean;
+  }
+
+  if(curl_url_set(ipfsurl, CURLUPART_PATH, pathbuffer, CURLU_URLENCODE)
+                  != CURLUE_OK) {
+    goto clean;
+  }
+
+  if(curl_url_get(ipfsurl, CURLUPART_URL, url, CURLU_URLENCODE)
+                  != CURLUE_OK) {
+    goto clean;
+  }
+
+  if(curl_url_get(ipfsurl, CURLUPART_SCHEME, &protocol, CURLU_DEFAULT_SCHEME)
+                  != CURLUE_OK) {
+    goto clean;
+  }
+
+  result = CURLE_OK;
+
+clean:
+  if(gateway) {
+    curl_free(gateway);
+  }
+  if(cid) {
+    curl_free(cid);
+  }
+  if(pathbuffer) {
+    curl_free(pathbuffer);
+  }
+  if(ipfsurl) {
+    curl_url_cleanup(ipfsurl);
+  }
+
+  return result;
+}
+
 /*
  * Return the protocol token for the scheme used in the given URL
  */
-static const char *url_proto(char *url)
+static const char *url_proto(char **url,
+                             struct OperationConfig *config,
+                             struct GlobalConfig *global)
 {
   CURLU *uh = curl_url();
   const char *proto = NULL;
 
   if(uh) {
-    if(url) {
-      if(!curl_url_set(uh, CURLUPART_URL, url,
-                       CURLU_GUESS_SCHEME | CURLU_NON_SUPPORT_SCHEME)) {
+    if(*url) {
+      if(curl_url_set(uh, CURLUPART_URL, *url,
+                      CURLU_GUESS_SCHEME | CURLU_NON_SUPPORT_SCHEME)
+         == CURLUE_OK) {
         char *schemep = NULL;
+
+        if(curl_url_get(uh, CURLUPART_SCHEME, &schemep,
+                         CURLU_DEFAULT_SCHEME) == CURLUE_OK
+           && schemep) {
+          if(curl_strnequal(schemep, "ipfs", 4)
+             || curl_strnequal(schemep, "ipns", 4)) {
```

Comment:
This would also match the schemes `ipfsXXXX://` and `ipnsYYYYY://`, right? I think this should use `curl_strequal()` instead.

Reply:
Yes, `curl_strequal` would be a better fit. Thanx, fixed.